snowflake/data_access_sync_from_target.go

```diff
@@ -0,0 +1,383 @@
+	//Get all database roles for each database and import them
+	databases, err := s.getDatabases(repo)
+	if err != nil {
+		return err
+	}
+
+	for _, database := range databases {
+		logger.Info(fmt.Sprintf("Reading roles from Snowflake inside database %s", database.Name))
+		processedAps := make(map[string]*exporter.AccessProvider)
+
+		// Get all database roles for database
+		roleEntities, err := repo.GetDatabaseRoles(database.Name)
+		if err != nil {
+			return err
+		}
+
+		for _, roleEntity := range roleEntities {
+			fullRoleName := fmt.Sprintf("%s.%s", database.Name, roleEntity.Name)
+			if _, exclude := excludedRoles[fullRoleName]; exclude {
+				logger.Info("Skipping SnowFlake DATABASE ROLE " + fullRoleName)
+				continue
+			}
+
+			err = s.importAccessForDatabaseRole(database.Name, roleEntity, externalGroupOwners, linkToExternalIdentityStoreGroups, repo, processedAps, shares)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = accessProviderHandler.AddAccessProviders(values(processedAps)...)
+		if err != nil {
+			return fmt.Errorf("error adding access provider to import file: %s", err.Error())
+		}
+	}
+
+	return nil
+}
+
+func (s *AccessSyncer) comesFromExternalIdentityStore(roleEntity RoleEntity, externalGroupOwners string) bool {
+	fromExternalIS := false
+
+	// check if Role Owner is part of the ones that should be (partially) locked
+	for _, i := range strings.Split(externalGroupOwners, ",") {
+		if strings.EqualFold(i, roleEntity.Owner) {
+			fromExternalIS = true
+		}
+	}
+
+	return fromExternalIS
+}
+
+func (s *AccessSyncer) importAccessForDatabaseRole(database string, roleEntity RoleEntity, externalGroupOwners string, linkToExternalIdentityStoreGroups bool, repo dataAccessRepository, processedAps map[string]*exporter.AccessProvider, shares []string) error {
+	logger.Info(fmt.Sprintf("Reading SnowFlake DATABASE ROLE %s inside %s", roleEntity.Name, database))
+
+	roleName := roleEntity.Name
+	externalId := databaseRoleExternalIdGenerator(database, roleName)
+	currentApType := ptr.String(apTypeDatabaseRole)
+	fromExternalIS := s.comesFromExternalIdentityStore(roleEntity, externalGroupOwners)
+
+	users, groups, accessProviders, err := s.retrieveWhoEntitiesForRole(roleEntity, externalId, currentApType, fromExternalIS, linkToExternalIdentityStoreGroups, repo)
+	if err != nil {
+		return err
+	}
+
+	ap, f := processedAps[externalId]
+	if !f {
+		processedAps[externalId] = &exporter.AccessProvider{
+			Type:       currentApType,
+			ExternalId: externalId,
+			// Updated this because of https://github.com/raito-io/appserver/blob/587484940a2e356a486dd8779166852761885353/lambda/appserver/services/access_provider/importer/importer.go#L523
+			ActualName: roleName,
+
+			Name:       roleName,
```

Comment:
Wondering if we should add the data source name to the (display) name of the AP to make it more distinguishable in the UI. Thoughts?

Reply:
Was indeed also one of my toughts, when using it currently in the UI you don't have any clue within what scope this is.